File: from-1601-to-1700/1642-furthest-building-you-can-reach/solution.py
import heapq

class Solution:
    def furthestBuilding(self, heights: List[int], bricks: int, ladders: int) -> int:
        heap = []

        bricks_count = 0

        for i in range(len(heights) - 1):
            diff = heights[i+1] - heights[i]
            if diff > 0:
                bricks_count += diff
                heapq.heappush(heap, -diff)

                if bricks_count > bricks:
                    if ladders > 0:
                        ladders -= 1
                        bricks_count += heapq.heappop(heap)
                    else:
                        return i
                    
        return len(heights) - 1

File: from-1601-to-1700/1642-furthest-building-you-can-reach/solution.go
func furthestBuilding(heights []int, bricks int, ladders int) int {
    h := &IntHeap{}
    heap.Init(h)
    for i := 1; i < len(heights); i++ {
        diff := heights[i] - heights[i - 1]
        if diff > 0 {
            heap.Push(h, diff)
            if h.Len() > ladders { bricks -= heap.Pop(h).(int) }
            if bricks < 0 { return i - 1 }
        }
    }
    return len(heights) - 1
}

type IntHeap []int

func (h IntHeap) Len() int           { return len(h) }
func (h IntHeap) Less(i, j int) bool { return h[i] < h[j] }
func (h IntHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }

func (h *IntHeap) Push(x interface{}) {
	*h = append(*h, x.(int))
}

func (h *IntHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}
